src/measure_theory/covering/besicovitch.lean

```diff
@@ -0,0 +1,440 @@
+guarantees that the Besicovitch covering theorem holds. It is satified by finite-dimensional
+real vector spaces. -/
+class has_besicovitch_covering (α : Type*) [metric_space α] : Prop :=
+(no_satellite_config : ∃ (N : ℕ) (τ : ℝ), 1 < τ ∧ is_empty (besicovitch.satellite_config α N τ))
+
+namespace besicovitch
+
+namespace satellite_config
+variables {α : Type*} [metric_space α] {N : ℕ} {τ : ℝ} (a : satellite_config α N τ)
+
+lemma inter' (i : fin N.succ) : dist (a.c i) (a.c (last N)) ≤ a.r i + a.r (last N) :=
+begin
+  rcases lt_or_le i (last N) with H|H,
+  { exact a.inter i H },
+  { have I : i = last N := top_le_iff.1 H,
+    have := (a.rpos (last N)).le,
+    simp only [I, add_nonneg this this, dist_self] }
+end
+
+lemma hlast' (i : fin N.succ) (h : 1 ≤ τ) : a.r (last N) ≤ τ * a.r i :=
+begin
+  rcases lt_or_le i (last N) with H|H,
+  { exact (a.hlast i H).2 },
+  { have : i = last N := top_le_iff.1 H,
+    rw this,
+    exact le_mul_of_one_le_left (a.rpos _).le h }
+end
+
+end satellite_config
+
+/-! ### Extracting disjoints subfamilies from a ball covering -/
+
+/-- A ball package is a family of balls in a metric space with positive bounded radii. -/
+structure ball_package (β : Type*) (α : Type*) :=
+(c : β → α)
+(r : β → ℝ)
+(rpos : ∀ b, 0 < r b)
+(r_bound : ℝ)
+(r_le : ∀ b, r b ≤ r_bound)
+
+/-- The ball package made of unit balls. -/
+def unit_ball_package (α : Type*) : ball_package α α :=
+{ c := id,
+  r := λ _, 1,
+  rpos := λ _, zero_lt_one,
+  r_bound := 1,
+  r_le := λ _, le_rfl }
+
+instance (α : Type*) : inhabited (ball_package α α) :=
+⟨unit_ball_package α⟩
+
+/-- A Besicovitch tau-package is a family of balls in a metric space with positive bounded radii,
+together with enough data to proceed with the Besicovitch greedy algorithm. We register this in
+a single structure to make sure that all our constructions in this algorithm only depend on
+one variable. -/
+structure tau_package (β : Type*) (α : Type*) extends ball_package β α :=
+(τ : ℝ)
+(one_lt_tau : 1 < τ)
+
+instance (α : Type*) : inhabited (tau_package α α) :=
+⟨{ τ := 2,
+  one_lt_tau := one_lt_two,
+  .. unit_ball_package α }⟩
+
+variables {α : Type*} [metric_space α] {β : Type u}
+
+namespace tau_package
+
+variables [nonempty β] (p : tau_package β α)
+include p
+
+/-- Choose inductively large balls with centers that are not contained in the union of already
+chosen balls. This is a transfinite induction. -/
+noncomputable def f : ordinal.{u} → β
+| i :=
+    -- `Z` is the set of points that are covered by already constructed balls
+    let Z := ⋃ (j : {j // j < i}), ball (p.c (f j)) (p.r (f j)),
+    -- `R` is the supremum of the radii of balls with centers not in `Z`
+    R := supr (λ b : {b : β // p.c b ∉ Z}, p.r b) in
+    -- return a color `b` for which the center `c b` is not in `Z`, and the radius is at
+    -- least `R / τ`, if such a color exists (and garbage otherwise).
```

Comment:
I'm not very familiar with this argument, but I find the usage of `color` confusing. I thought that the coloring was a map to `ℕ` defined below. Does it make sense to say that it returns a "ball `b` from the package, for which..." ?

Reply:
Search and replace problem on my side. Initially I used `index` instead of `color` in all the file, then I realized that `color` would be better so I did a global replace in the file, but here it should have remained `index`.